controllers/appgroup_controller_test.go

```diff
@@ -1,556 +1,430 @@
+				return false
+			}
+			if err := k8sClient.Get(ctx, key, appGroup); err != nil {
+				return false
+			}
+			return hr.Spec.Chart.Spec.Version == ambassadorChartVersion && appGroup.GetReadyCondition() == meta.SucceededReason
+		}, DefaultTimeout, time.Second).Should(BeTrue())
+
+	})
+
+	It("Should succeed to rollback helm chart versions on failure", func() {
+		appGroup.Spec.Applications[1].Spec.Chart.Version = ambassadorOldChartVersion
+
+		By("Applying the bookinfo object to the cluster")
+		err = k8sClient.Create(ctx, appGroup)
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Waiting for the bookinfo object to reach a succeeded reason")
+		Eventually(func() bool {
+			if err := k8sClient.Get(ctx, key, appGroup); err != nil {
+				return false
+			}
+			_, exist := appGroup.Annotations[v1alpha1.LastSuccessfulAnnotation]
+			return appGroup.GetReadyCondition() == meta.SucceededReason && exist
+		}, DefaultTimeout, time.Second).Should(BeTrue())
+
+		By("Upgrading the ambassador chart to a newer version while intentionally timing out the last DAG step")
+		patch := client.MergeFrom(appGroup.DeepCopy())
+		appGroup.Spec.Applications[1].Spec.Chart.Version = ambassadorChartVersion
+		appGroup.Spec.Applications[0].Spec.Release.Timeout = &metav1.Duration{Duration: time.Second}
+		err = k8sClient.Patch(ctx, appGroup, patch)
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Waiting for the application group to start its upgrade")
+		Eventually(func() bool {
+			if err := k8sClient.Get(ctx, key, appGroup); err != nil {
+				return false
+			}
+			return appGroup.GetReadyCondition() == meta.ProgressingReason && appGroup.Generation > 1
+		}, time.Second*30, time.Second).Should(BeTrue())
+
+		By("Waiting for the newer version of the charts to be released")
+		Eventually(func() bool {
+			hr := &fluxhelmv2beta1.HelmRelease{}
+			hrKey := objKeyBuilder(ambassador, name)
+			if err := k8sClient.Get(ctx, hrKey, hr); err != nil {
+				return false
+			}
+			return hr.Spec.Chart.Spec.Version == ambassadorChartVersion && meta.GetResourceCondition(hr, meta.ReadyCondition).Reason == meta2.ReconciliationSucceededReason
+		}, DefaultTimeout, time.Second).Should(BeTrue())
+
+		By("Ensuring that the applications rollback to their starting version")
+		Eventually(func() bool {
+			hr := &fluxhelmv2beta1.HelmRelease{}
+			hrKey := objKeyBuilder(ambassador, name)
+			if err := k8sClient.Get(ctx, hrKey, hr); err != nil {
+				return false
+			}
+			if err := k8sClient.Get(ctx, key, appGroup); err != nil {
+				return false
+			}
+			return hr.Spec.Chart.Spec.Version == ambassadorOldChartVersion && meta.GetResourceCondition(hr, meta.ReadyCondition).Reason == meta2.ReconciliationSucceededReason && appGroup.GetReadyCondition() == meta.WorkflowFailedReason && appGroup.GetWorkflowCondition(v1alpha1.Rollback) == meta.SucceededReason
+		}, DefaultTimeout, time.Second).Should(BeTrue())
+	})
+
+	It("Should create the bookinfo and then delete it while in progress", func() {
```

Comment:
Is there a similar test for update while in progress?

Reply:
there isn't